Bneukom/improve bp api

bneukom opened the pull request:
- Added Blueprint support to:
  - Change the Initial Shape (Spline or StaticMesh) of a VitruvioComponent
  - Convert a StaticMeshActor to a VitruvioActor
  - Set array attribute on a VitruvioComponent
- Added async support of VitruvioComponent modifications. The Blueprint functions have callbacks when the generate/rule evaluation has completed (eg when changing an attribute). This is helpful to properly chain together several Blueprint calls.
- Added a copy fully qualified attribute name to the details panel
- Fixed some minor issues and typos

chr11115 commented on VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Public/GenerateCompletedCallbackProxy.h:
```diff
@@ -32,15 +32,19 @@ class VITRUVIO_API UGenerateCompletedCallbackProxy final : public UBlueprintAsyn
 	/** Called after the attributes have been evaluated. Note that it is not guaranteed that this callback is ever called. */
 	UPROPERTY(BlueprintAssignable, meta = (DisplayName = "Attributes Evaluated"), Category = "Vitruvio")
 	FOnAttributesEvaluatedDelegate OnAttributesEvaluated;
+	FOnAttributesEvaluatedDelegateCpp OnAttributesEvaluatedCpp;
```
We could probably use a more descriptive name here

bneukom replied:
Ah yes typically we actually do the opposite. OnAttributesEvaluatedBlueprint and then using the DisplayName to override a nice name for the Blueprints. Will change that.

chr11115 commented on VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Public/GenerateCompletedCallbackProxy.h:
```diff
@@ -32,15 +32,19 @@ class VITRUVIO_API UGenerateCompletedCallbackProxy final : public UBlueprintAsyn
 	/** Called after generate has completed. Note that it is not guaranteed that this callback is ever called. */
 	UPROPERTY(BlueprintAssignable, meta = (DisplayName = "Generate Completed"), Category = "Vitruvio")
 	FGenerateCompletedDelegate OnGenerateCompleted;
+	FGenerateCompletedDelegateCpp OnGenerateCompletedCpp;
```
See above

bneukom replied:
See above ;)

chr11115 commented on VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Public/VitruvioBlueprintLibrary.h:
```diff
@@ -0,0 +1,50 @@
+
+	/**
+	 * Returns whether the given Actor can be converted to a VitruvioActor (see also ConvertToVitruvioActor). Converting an Actor to a VitruvioActor
+	 * is only possible if the Actor already has a valid initial shape component attached (eg. a StaticMeshComponent).
```
We should probably be more precise as to what kind of actors can be converted. (I.e. we can't convert a regular actor with vitruvio component attached to a VitruvioActor)

bneukom replied:
Hm I thought the second sentence explained that? 

"Converting an Actor to a VitruvioActor is only possible if the Actor already has a valid initial shape component attached".

Maybe by adding something like "and does not already have a VitruvioComponent attached"?

chr11115 replied:
Sounds good. I was just confused as I thought we could "promote" a regular actor that used a vitruvio component to a proper vitruvio actor. I think the changes you propose would definitely clarify this confusion.

bneukom replied:
Ahh ok! Makes sense then.

chr11115 reviewed (changes requested):
Overall nice work!
Some minor clarification nags and improvement proposals. 
Also another minor thing, but that's a style question is that I find if(ptr != nullptr) a little more readable for nullptr checks compared to if(ptr).

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/GenerateCompletedCallbackProxy.cpp:
```diff
@@ -0,0 +1,205 @@
+}
+
+UGenerateCompletedCallbackProxy* UGenerateCompletedCallbackProxy::ConvertToVitruvioActor(UObject* WorldContextObject, const TArray<AActor*>& Actors,
+																						 TArray<AVitruvioActor*>& OutVitruvioActors,
```
The last two parameters have default values so we can't move it.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioComponent.cpp:
```diff
@@ -75,29 +76,42 @@ bool GetAttribute(const TMap<FString, URuleAttribute*>& Attributes, const FStrin
+			return;
+		}
+
+		// Create new attribute if it does not exist and bAddIfNonExisting is true
```
Very good point. It was quite confusing to always have to wait for an attribute evaluation callback to properly set the attributes. I made a change that attributes are evaluated before setting them if they have not been evaluated.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioComponent.cpp:
```diff
@@ -266,6 +266,37 @@ bool UVitruvioComponent::GetFloatAttribute(const FString& Name, float& OutValue)
+void UVitruvioComponent::SetAttributes(const TMap<FString, FString>& NewAttributes)
+{
+	const bool bOldGenerateAutomatically = GenerateAutomatically;
+	GenerateAutomatically = false;
```
Also good point. All the BP API functions now have an additional flag if a model should be generated (always defaults to true). The GenerateAutomatically is still used for changes to a VitruvioActor through the UI.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/GenerateCompletedCallbackProxy.cpp:
```diff
@@ -0,0 +1,134 @@
+			AVitruvioActor* VitruvioActor = Actor->GetWorld()->SpawnActor<AVitruvioActor>(Actor->GetActorLocation(), Actor->GetActorRotation());
+
+			USceneComponent* NewInitialShapeSceneComponent = CopyInitialShapeSceneComponent(Actor, VitruvioActor);
+			NewInitialShapeSceneComponent->SetWorldTransform(VitruvioActor->GetTransform());
```
Applying it to the Actor and the root component is the same (SetTransform on the Actor just forwards it to its root component). But I did cleanup the code a bit anyways (eg setting the "InitialShapeComponent" as the root component of the new Actor and also setting the transform internally)

bneukom commented:
Implemented most of the requested changes:
- Evaluate attributes before all SetXYZAttribute calls
- Removed the addIfNonExisting flags to SetXYZAttribute functions
- Added a GenerateModel flag to all BP API functions. GenerateAutomatically is only used for changes made through the UI

Also added support to add array attribute values to the genertic SetAttributes call.

chr11115 reviewed (changes requested):
Nice Work! Almost Ready.. 
When setting the attributes with the BeginPlay event it currently does not generate the building, because The vitruvioActor is late initialized and overrides any previous generate/setAttributes call..

bneukom commented:
Thanks @chr11115 I fixed that and merged it!
